File: mcp-authorization/fastmcp-python/scripts/verify.py
#!/usr/bin/env python3
"""Check the four things an MCP client checks, before you plug in a real client.

Run the server, then ``python scripts/verify.py``. Every line is one protocol
requirement, so a failure points at the exact step that is misconfigured.

    MCP_BASE_URL   where this process listens (default http://localhost:3005)
    NAMOID_ISSUER / NAMOID_MCP_RESOURCE   read from .env if not already set
"""

from __future__ import annotations

import os
import re
import sys
from pathlib import Path
from urllib.parse import urlsplit

import httpx

failures = 0


def load_dotenv(path: Path) -> None:
    if not path.is_file():
        return
    for line in path.read_text().splitlines():
        match = re.match(r"^\s*([A-Z0-9_]+)\s*=\s*(.*)$", line)
        if not match:
            continue
        key, raw_value = match.groups()
        os.environ.setdefault(key, raw_value.strip().strip("\"'"))


def require_env(name: str) -> str:
    value = (os.environ.get(name) or "").strip()
    if not value:
        sys.exit(f"{name} is required. Copy .env.example to .env and fill it in.")
    return value


def check(label: str, run) -> None:
    global failures
    try:
        print(f"  ok    {label}\n        {run()}")
    except Exception as exc:  # noqa: BLE001 - a failed check is reported, not raised
        failures += 1
        print(f"  FAIL  {label}\n        {exc}")


def metadata_path(resource: str) -> str:
    """RFC 9728 inserts the well-known segment between authority and path."""
    path = urlsplit(resource).path
    suffix = "" if path in ("", "/") else path.rstrip("/")
    return f"/.well-known/oauth-protected-resource{suffix}"


def post_mcp(token: str | None = None) -> httpx.Response:
    headers = {
        "content-type": "application/json",
        "accept": "application/json, text/event-stream",
    }
    if token:
        headers["authorization"] = f"Bearer {token}"
    return httpx.post(
        f"{BASE_URL}{MCP_PATH}",
        headers=headers,
        json={"jsonrpc": "2.0", "id": 1, "method": "tools/list"},
        timeout=10.0,
    )


load_dotenv(Path(__file__).resolve().parent.parent / ".env")

BASE_URL = (os.environ.get("MCP_BASE_URL") or "http://localhost:3005").rstrip("/")
ISSUER = require_env("NAMOID_ISSUER").rstrip("/")
RESOURCE = require_env("NAMOID_MCP_RESOURCE")
MCP_PATH = urlsplit(RESOURCE).path or "/"
METADATA_PATH = metadata_path(RESOURCE)


def check_server_running() -> str:
    # FastMCP has no unauthenticated health route, so the metadata document
    # doubles as the liveness probe.
    response = httpx.get(f"{BASE_URL}{METADATA_PATH}", timeout=10.0)
    if response.status_code != 200:
        raise RuntimeError(f"GET {METADATA_PATH} returned {response.status_code}")
    return BASE_URL


def check_discovery() -> str:
    url = f"{ISSUER}/.well-known/oauth-authorization-server"
    response = httpx.get(url, headers={"accept": "application/json"}, timeout=10.0)
    if response.status_code != 200:
        raise RuntimeError(f"GET {url} returned {response.status_code}")
    document = response.json()
    if document.get("issuer") != ISSUER:
        raise RuntimeError(f"issuer mismatch: expected {ISSUER}, document says {document.get('issuer')}")
    pkce = ",".join(document.get("code_challenge_methods_supported") or [])
    cimd = "yes" if document.get("client_id_metadata_document_supported") is True else "no"
    return f"{document['issuer']} (PKCE {pkce}, CIMD {cimd})"


def check_protected_resource_metadata() -> str:
    response = httpx.get(f"{BASE_URL}{METADATA_PATH}", timeout=10.0)
    if response.status_code != 200:
        raise RuntimeError(f"GET {METADATA_PATH} returned {response.status_code}")
    document = response.json()
    if document.get("resource") != RESOURCE:
        raise RuntimeError(f"metadata resource is {document.get('resource')}, expected {RESOURCE}")
    servers = document.get("authorization_servers") or []
    if ISSUER not in [str(server).rstrip("/") for server in servers]:
        raise RuntimeError(f"metadata does not list {ISSUER} as an authorization server")
    scopes = " ".join(document.get("scopes_supported") or [])
    return f"{document['resource']} -> {', '.join(str(s) for s in servers)} [{scopes}]"


def check_unauthenticated_challenge() -> str:
    response = post_mcp()
    if response.status_code != 401:
        raise RuntimeError(f"expected 401, received {response.status_code}")
    challenge = response.headers.get("www-authenticate", "")
    if not challenge.lower().startswith("bearer"):
        raise RuntimeError("WWW-Authenticate is missing or is not a Bearer challenge")
    if "resource_metadata=" not in challenge:
        raise RuntimeError("WWW-Authenticate does not point at resource_metadata")
    return challenge


def check_forged_token_rejected() -> str:
    response = post_mcp("not-a-real-token")
    if response.status_code != 401:
        raise RuntimeError(f"expected 401, received {response.status_code}")
    return response.headers.get("www-authenticate", "401 without a challenge header")


check("server is running", check_server_running)
check("NamoID discovery is reachable", check_discovery)
check("protected-resource metadata is published", check_protected_resource_metadata)
check("unauthenticated call returns a discovery challenge", check_unauthenticated_challenge)
check("a forged token is rejected", check_forged_token_rejected)

print("")
if failures:
    print(f"{failures} check(s) failed.")
    sys.exit(1)
print("All checks passed. The resource-server half of the flow is correct.")
print(f"Connect a client at {RESOURCE} using a preregistered Client ID or a")
print('Client ID Metadata Document — see "Connect an MCP client" in the README.')

File: mcp-authorization/express-ts/scripts/verify.mjs
#!/usr/bin/env node
/**
 * Check the four things an MCP client checks, before you plug in a real client.
 *
 * Run the server, then `pnpm verify`. Every line is one protocol requirement,
 * so a failure points at the exact step that is misconfigured.
 *
 *   MCP_BASE_URL   where this process listens (default http://localhost:3004)
 *   NAMOID_ISSUER / NAMOID_MCP_RESOURCE   read from .env if not already set
 */

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

loadDotEnv(new URL("../.env", import.meta.url));

const baseUrl = trimSlash(process.env.MCP_BASE_URL ?? "http://localhost:3004");
const issuer = trimSlash(requireEnv("NAMOID_ISSUER"));
const resource = requireEnv("NAMOID_MCP_RESOURCE");
const metadataPath = protectedResourceMetadataPath(resource);

let failures = 0;

await check("server is running", async () => {
  const response = await fetch(`${baseUrl}/health`);
  if (!response.ok) throw new Error(`GET /health returned ${response.status}`);
  return baseUrl;
});

await check("NamoID discovery is reachable", async () => {
  const url = `${issuer}/.well-known/oauth-authorization-server`;
  const response = await fetch(url, { headers: { accept: "application/json" } });
  if (!response.ok) throw new Error(`GET ${url} returned ${response.status}`);
  const document = await response.json();
  if (document.issuer !== issuer) {
    throw new Error(`issuer mismatch: expected ${issuer}, document says ${document.issuer}`);
  }
  const cimd = document.client_id_metadata_document_supported === true;
  return `${document.issuer} (PKCE ${(document.code_challenge_methods_supported ?? []).join(",")}, CIMD ${cimd ? "yes" : "no"})`;
});

await check("protected-resource metadata is published", async () => {
  const response = await fetch(`${baseUrl}${metadataPath}`);
  if (!response.ok) throw new Error(`GET ${metadataPath} returned ${response.status}`);
  const document = await response.json();
  if (document.resource !== resource) {
    throw new Error(`metadata resource is ${document.resource}, expected ${resource}`);
  }
  if (!(document.authorization_servers ?? []).includes(issuer)) {
    throw new Error(`metadata does not list ${issuer} as an authorization server`);
  }
  return `${document.resource} -> ${document.authorization_servers.join(", ")} [${(document.scopes_supported ?? []).join(" ")}]`;
});

await check("unauthenticated call returns a discovery challenge", async () => {
  const response = await postMcp();
  if (response.status !== 401) throw new Error(`expected 401, received ${response.status}`);
  const challenge = response.headers.get("www-authenticate") ?? "";
  if (!challenge.toLowerCase().startsWith("bearer")) {
    throw new Error("WWW-Authenticate is missing or is not a Bearer challenge");
  }
  if (!challenge.includes("resource_metadata=")) {
    throw new Error("WWW-Authenticate does not point at resource_metadata");
  }
  return challenge;
});

await check("a forged token is rejected", async () => {
  const response = await postMcp("not-a-real-token");
  if (response.status !== 401) throw new Error(`expected 401, received ${response.status}`);
  const body = await response.json().catch(() => ({}));
  if (body.error !== "invalid_token") {
    throw new Error(`expected error "invalid_token", received ${JSON.stringify(body.error)}`);
  }
  return "invalid_token";
});

console.log("");
if (failures > 0) {
  console.log(`${failures} check(s) failed.`);
  process.exit(1);
}
console.log("All checks passed. The resource-server half of the flow is correct.");
console.log(`Connect a client at ${resource} using a preregistered Client ID or a`);
console.log("Client ID Metadata Document — see \"Connect an MCP client\" in the README.");

function postMcp(token) {
  return fetch(`${baseUrl}/mcp`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      accept: "application/json, text/event-stream",
      ...(token ? { authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({ jsonrpc: "2.0", id: 1, method: "tools/list" }),
  });
}

async function check(label, run) {
  try {
    const detail = await run();
    console.log(`  ok    ${label}\n        ${detail}`);
  } catch (error) {
    failures += 1;
    console.log(`  FAIL  ${label}\n        ${error instanceof Error ? error.message : error}`);
  }
}

/** RFC 9728 inserts the well-known segment between authority and resource path. */
function protectedResourceMetadataPath(resourceUrl) {
  const { pathname } = new URL(resourceUrl);
  return `/.well-known/oauth-protected-resource${pathname === "/" ? "" : pathname}`;
}

function loadDotEnv(url) {
  let contents;
  try {
    contents = readFileSync(fileURLToPath(url), "utf8");
  } catch {
    return;
  }
  for (const line of contents.split("\n")) {
    const match = /^\s*([A-Z0-9_]+)\s*=\s*(.*)$/.exec(line);
    if (!match) continue;
    const [, key, rawValue] = match;
    if (process.env[key] !== undefined) continue;
    process.env[key] = rawValue.trim().replace(/^["']|["']$/g, "");
  }
}

function requireEnv(name) {
  const value = process.env[name]?.trim();
  if (!value) {
    console.error(`${name} is required. Copy .env.example to .env and fill it in.`);
    process.exit(1);
  }
  return value;
}

function trimSlash(value) {
  return value.replace(/\/+$/, "");
}
